### PR TITLE
[SPARK-46064][SQL][SS] Move out EliminateEventTimeWatermark to the analyzer and change to only take effect on resolved child

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to move out EliminateEventTimeWatermark to the analyzer (one of the analysis rule), and also make a change to eliminate EventTimeWatermark node only when the child of EventTimeWatermark is "resolved".

### Why are the changes needed?

Currently, we apply EliminateEventTimeWatermark immediately when withWatermark is called, which means the rule is applied immediately against the child, regardless whether child is resolved or not.

It is not an issue for the usage of DataFrame API initiated by read / readStream, because streaming sources have the flag isStreaming set to true even it is yet resolved. But mix-up of SQL and DataFrame API would expose the issue; we may not know the exact value of isStreaming flag on unresolved node and it is subject to change upon resolution.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UTs.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan PTAL, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -347,7 +347,9 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     Batch("Cleanup", fixedPoint,
       CleanupAliases),
     Batch("HandleSpecialCommand", Once,
-      HandleSpecialCommand)
+      HandleSpecialCommand),
+    Batch("Remove watermark for batch query", Once,
```
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:351]: I guess the node is expected to be already resolved from this point, but I'm checking the flag of resolved in EliminateEventTimeWatermark to make the logic be self-describe.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TestRelations.scala [hunk 2]
```diff
@@ -68,4 +68,18 @@ object TestRelations {
 
   val mapRelation = LocalRelation(
     AttributeReference("map", MapType(IntegerType, IntegerType))())
+
+  val streamingRelation = LocalRelation(
```
Comment (HeartSaVioR) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TestRelations.scala:72]: This is also added in #43966 - I'll rebase when either one is merged.

**Review by cloud-fan — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
https://github.com/HeartSaVioR/spark/actions/runs/6967344462/job/18959281391

failures do not relate to the change. SQL tests are all passing.

**Comment by HeartSaVioR** [comment]
Thanks, merging to master/3.5/3.4.

**Pull request closed without merging**